keps/0007-dynamic-audit-configuration.md

```diff
@@ -0,0 +1,231 @@
+
+    // ThrottleEnabled determines whether throttling is enabled
+    ThrottleEnabled *bool `json:"throttleEnabled,omitempty"`
+
+    // ThrottleQPS maximum number of batches per second
+    ThrottleQPS *float32 `json:"thottleQPS,omitempty"`
+}
+
+// WebhookClientConfig contains the information to make a TLS
+// connection with the webhook; this follows: 
+// https://github.com/kubernetes/api/blob/master/admissionregistration/v1beta1/types.go#L222
+// but may require some additive auth parameters
+type WebhookClientConfig struct {
+    // URL of the server
+    URL *string `json:"url,omitempty"`
+
+    // Service name to send to
+    Service *ServiceReference `json:"service,omitempty"`
+
+    // `caBundle` is a PEM encoded CA bundle which will be used to validate
+    // the webhook's server certificate.
+    CABundle []byte `json:"caBundle"`
+}
+```
+
+This configuration object will be accepted both as the existing runtime flags, and a dynamic api object. Multiple definitions can exist as independent solutions. These updates will require the audit API to be registered with the apiserver. Dynamic configuration will be enabled by a feature gate. If existing flags are provided to configure the audit backend they will be taken as a separate configuration object.
+
+Example cluster yaml config:   
+```yaml
+apiVersion: audit.k8s.io/v1beta1
+kind: ClusterAuditConfiguration
+metadata:
+  name: <name>
+backends:
+  webhooks:
+  - mode: <batch>
+    initialBackoff: <10s>
+    batch:
+      bufferSize: <400>
+      maxSize: <10000>
+      maxWait: <30s>
+      throttleBurst: <15>
+      throttleEnabled: <true>
+      throttleQPS: <10>
+    clientConfig:
+      url: <backend url>
+      service: <optional service name>
+      caBundle: <ca bundle>
+```
+
+### User Stories
+
+#### Story 1
+As a cluster admin, I will easily be able to enable the interal auditing features of an existing cluster, and tweak the configurations as necessary. I want to prevent privilege escalation from being able to tamper with a root audit configuration.
+
+#### Story 2
+As a Kubernetes extension developer, I will be able to provide drop in extensions that utilize audit data.
+
+#### Story 3
+As a cluster admin, I will be able configure multiple audit-policies and webhook endpoints to provide independent auditing facilities.
+
+### Implementation Details/Notes/Constraints
+
+When an update to a configuration is issued, the server will cache the current working configuration. If the update fails to apply, the system will roll back to the previous configuration and log that action, if configured through the api a 400 will be returned. If no previous configuration exists the application of the config will fail, and log the result. Any actions to the audit configuration objects will be hard coded to log at the `level=RequestResponse` to the previous backend and the new backend. If multiple API servers exist, the configuration will be rolled out in increments.
```

Comment:
Sorry, I still don't understand this. What specifically does this mean?

> If the update fails to apply, ...

Maybe it would help if you describe the implementation of that condition a bit more?

Reply:
I think this only applies to policy, since were not handling that here I'll remove